src/main/java/frc/team2412/robot/subsystems/LimelightSubsystem.java

```diff
@@ -0,0 +1,196 @@
+		limelightTab
+				.addDouble("Horizontal Offset", this::getHorizontalOffset)
+				.withPosition(1, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Vertical Offset", this::getVerticalOffset)
+				.withPosition(2, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Target Distance 2 - TEST ", this::getDistanceFromTarget)
+				.withPosition(4, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addDouble("Limelight Based Turn Power - TEST ", this::turnPowerLin)
+				.withPosition(5, 0)
+				.withSize(1, 1);
+		limelightTab
+				.addString("Current Pose ", this::getCurrentPoseString)
+				.withPosition(0, 1)
+				.withSize(4, 1);
+
+		limelightTab
+				.addString("Target Pose ", this::getTargetPoseString)
+				.withPosition(0, 2)
+				.withSize(4, 1);
+	}
+
+	private void setPipeline() {
+		networkTable.getEntry("pipeline").setNumber(2);
+	}
+
+	// METHODS
+
+	public boolean hasTargets() {
+		return (networkTable.getEntry("tv").getDouble(0) != 0);
+	}
+
+	public double getHorizontalOffset() {
+		return networkTable.getEntry("tx").getDouble(0);
+	}
+
+	public double getVerticalOffset() {
+		return networkTable.getEntry("ty").getDouble(0);
+	}
+
+	public double getBoxWidth() {
+		return networkTable.getEntry("tshort").getDouble(0);
+	}
+	//TODO re-measure distances to make this right
+	public double getDistanceFromTarget() {
+
+		// focal length = (P x D) / W
+		double focal_length = 346.1818;
+
+		// distance = (W x F) / P
+		// returns inches for testing purposes, will divide by 39.3700787 to return meters
+		return (8.25 * focal_length) / getBoxWidth();
+	}
+
+	// returns turn power from exponential curve, input is from -27deg to +27deg
+	public double turnPowerExp() {
+		if (Math.abs(getHorizontalOffset()) >= 1.0) {
+			return MathUtil.clamp(
+					(Math.signum(getHorizontalOffset())
+							* (Math.pow(Math.abs(getHorizontalOffset()), 0.6) * 0.14)),
+					-0.25,
+					0.25);
+		} else {
+			return 0.0;
+		}
+	}
+
+	// returns turn power from linear curve, input is from -27deg to +27deg
+	public double turnPowerLin() {
+		if (Math.abs(getHorizontalOffset()) >= 1.0) {
+			return MathUtil.clamp(0.04 * getHorizontalOffset(), -0.25, 0.25);
+		} else {
+			return 0.0;
+		}
+	}
+
+	// tan(degree) * distance = sideways distance
+
+	// target height / tan(vertical angle)
+
+	public Pose2d getTargetPose(Pose2d currentPose) {
+
+		// math thing to get target pose using current pose
+
+		Rotation2d targetHeading =
+				new Rotation2d(
+						currentPose.getRotation().getRadians() + Units.degreesToRadians(getHorizontalOffset()));
+		double targetDistance = getDistanceFromTarget() / 39.3700787;
+
+		double targetX = Math.sin(targetHeading.getRadians()) * targetDistance;
+		double targetY = Math.cos(targetHeading.getRadians()) * targetDistance;
+
+		Pose2d targetPose =
+				new Pose2d(currentPose.getX() + targetY, currentPose.getY() + targetX, targetHeading);
```

Comment:
```suggestion
        Translation2d translationOffset = new Translation2d(targetDistance, targetHeading);

		Pose2d targetPose =
				new Pose2d(currentPose.getTranslation().plus(translationOffset), targetHeading);
```
The variables had the wrong name earlier (`targetX` and `targetY` were swapped), and this also more concisely expresses the logic.

Reply:
(The suggested change was undone in ae0af- I'm guessing the device that pushed that commit hadn't pulled in the commit from the suggestion)

Reply:
re-added